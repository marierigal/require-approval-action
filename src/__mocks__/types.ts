import { REVIEW_STATE } from '../constants';
import {
  DeleteWorkflowRunParams,
  ListReviewsParams,
  ListWorkflowRunsParams,
} from '../types';

export interface MockCore {
  __setMockInputs: (inputs: Record<string, string>) => void;
  getInput: (name: string) => string | undefined;
}

export interface MockGithub {
  __setMockContext: (pull_number?: number) => MockContext;
  __makeMockReview: (state: REVIEW_STATE, userId: number) => MockReview;
  __setMockReviews: (reviews: MockReview[]) => void;
  __makeMockRun: (runId: number, pull_number: number) => MockRun;
  __setMockRuns: (runs: MockRun[]) => void;
  context: MockContext;
  getOctokit: (token: string) => MockOctokit;
}

export interface MockOctokit {
  pulls: {
    listReviews: (params: ListReviewsParams) => Promise<{ data: MockReview[] }>;
  };
  actions: {
    listWorkflowRuns: (
      params: ListWorkflowRunsParams
    ) => Promise<{ data: { workflow_runs: MockRun[] } }>;
    deleteWorkflowRun: (params: DeleteWorkflowRunParams) => Promise<void>;
  };
}

export interface MockContext {
  eventName: string;
  workflow_id: number;
  repo: {
    owner: string;
    repo: string;
  };
  payload: {
    pull_request?: { number: number };
  };
}

export interface MockReview {
  user: { id: number };
  state: REVIEW_STATE;
}

export interface MockRun {
  id: number;
  pull_requests: { number: number }[];
}
