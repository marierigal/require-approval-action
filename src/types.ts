import { RestEndpointMethodTypes } from '@octokit/rest';

export type ListReviewsParams = RestEndpointMethodTypes['pulls']['listReviews']['parameters'];
export type Review = RestEndpointMethodTypes['pulls']['getReview']['response']['data'];
export type SortedReviews = { approved: Review[]; changesRequested: Review[] };

export type ListWorkflowRunsParams = RestEndpointMethodTypes['actions']['listWorkflowRuns']['parameters'];
export type DeleteWorkflowRunParams = RestEndpointMethodTypes['actions']['deleteWorkflowRun']['parameters'];
export type Run = RestEndpointMethodTypes['actions']['getWorkflowRun']['response']['data'];
