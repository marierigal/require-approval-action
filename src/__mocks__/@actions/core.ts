import { MockCore } from '../types';

const core = jest.createMockFromModule<MockCore>('@actions/core');

let mockInputs: Record<string, string> = {
  token: 'token',
  approvals: '3',
};

function __setMockInputs(inputs: Record<string, string>): void {
  mockInputs = inputs;
}

function getInput(name: string): string | undefined {
  return mockInputs[name];
}

core.__setMockInputs = __setMockInputs;
core.getInput = getInput;

module.exports = core;
