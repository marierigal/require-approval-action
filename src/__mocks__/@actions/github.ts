import { REVIEW_STATE } from '../../constants';
import {
  DeleteWorkflowRunParams,
  ListReviewsParams,
  ListWorkflowRunsParams,
} from '../../types';
import {
  MockContext,
  MockGithub,
  MockOctokit,
  MockReview,
  MockRun,
} from '../types';

const github = jest.createMockFromModule<MockGithub>('@actions/github');

const context: MockContext = {
  eventName: 'pull_request_review',
  workflow_id: 1,
  repo: {
    owner: 'owner',
    repo: 'repo',
  },
  payload: {},
};
let mockReviews: MockReview[] = [];
let mockRuns: MockRun[] = [];

function __setMockContext(pull_number?: number): MockContext {
  if (pull_number) {
    context.payload = {
      pull_request: { number: pull_number },
    };
  }

  return context;
}

function __makeMockReview(state: REVIEW_STATE, userId: number): MockReview {
  return {
    user: { id: userId },
    state: state,
  };
}

function __setMockReviews(reviews: MockReview[]): void {
  mockReviews = reviews;
}

function __makeMockRun(runId: number, pull_number: number): MockRun {
  return {
    id: runId,
    pull_requests: [{ number: pull_number }],
  };
}

function __setMockRuns(runs: MockRun[]): void {
  mockRuns = runs;
}

function listReviews(
  _params: ListReviewsParams
): Promise<{ data: MockReview[] }> {
  return new Promise(resolve => resolve({ data: mockReviews }));
}

function listWorkflowRuns(
  _params: ListWorkflowRunsParams
): Promise<{ data: { workflow_runs: MockRun[] } }> {
  return new Promise(resolve =>
    resolve({
      data: { workflow_runs: mockRuns },
    })
  );
}

function deleteWorkflowRun({ run_id }: DeleteWorkflowRunParams): Promise<void> {
  return new Promise(resolve => {
    mockRuns = mockRuns.filter(run => run.id !== run_id);
    resolve();
  });
}

function getOctokit(_token: string): MockOctokit {
  return {
    pulls: {
      listReviews,
    },
    actions: {
      listWorkflowRuns,
      deleteWorkflowRun,
    },
  };
}

github.__setMockContext = __setMockContext;
github.__makeMockReview = __makeMockReview;
github.__setMockReviews = __setMockReviews;
github.__makeMockRun = __makeMockRun;
github.__setMockRuns = __setMockRuns;
github.context = context;
github.getOctokit = getOctokit;

module.exports = github;
