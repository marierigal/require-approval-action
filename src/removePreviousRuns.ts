import { getOctokit } from '@actions/github';
import { ListWorkflowRunsParams } from './types';

export async function removePreviousRuns(
  token: string,
  params: ListWorkflowRunsParams,
  pull_number: number
): Promise<void> {
  const octokit = getOctokit(token);

  const {
    data: { workflow_runs },
  } = await octokit.actions.listWorkflowRuns(params);

  const runs = workflow_runs.filter(
    run => run.pull_requests[0]?.number === pull_number
  );

  for (const run of runs) {
    await octokit.actions.deleteWorkflowRun({
      ...params,
      run_id: run.id,
    });
  }
}
