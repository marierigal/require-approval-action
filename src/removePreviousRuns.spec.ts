import * as github from '@actions/github';
import { MockGithub } from './__mocks__/types';
import { removePreviousRuns } from './removePreviousRuns';

jest.mock('@actions/github');

const mockGithub = (github as unknown) as MockGithub;

describe('removePreviousRuns', () => {
  const token = 'token';
  const params = {
    owner: 'owner',
    repo: 'repo',
    workflow_id: 1,
    event: 'event',
  };
  const pull_number = 1;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should remove previous runs', async () => {
    mockGithub.__setMockContext(1);
    mockGithub.__setMockRuns([
      mockGithub.__makeMockRun(1, 1),
      mockGithub.__makeMockRun(2, 1),
      mockGithub.__makeMockRun(3, 2),
    ]);

    await removePreviousRuns(token, params, pull_number);

    return expect(
      mockGithub.getOctokit('token').actions.listWorkflowRuns(params)
    ).resolves.toMatchSnapshot();
  });
});
