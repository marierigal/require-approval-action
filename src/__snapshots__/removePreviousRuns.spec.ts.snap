// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removePreviousRuns should remove previous runs 1`] = `
Object {
  "data": Object {
    "workflow_runs": Array [
      Object {
        "id": 3,
        "pull_requests": Array [
          Object {
            "number": 2,
          },
        ],
      },
    ],
  },
}
`;
